### DISCUSSION TITLE
Documentation of the Files and File objects?

### DISCUSSION THREAD

**Discussion by fralau** [mkdocs/mkdocs#2375] [Q&A]
The question was raised within an issue of [mkdocs-macros](https://github.com/fralau/mkdocs_macros_plugin): [what is the structure of the Files and File objects, and how could those objects be used programmatically](https://github.com/fralau/mkdocs_macros_plugin/issues/80#issuecomment-822988190)?

There is little documentation available on those two objects, except about events, [to say that `files` is a collection.](https://www.mkdocs.org/user-guide/plugins/#events).

Aside of [the original Python module](https://github.com/mkdocs/mkdocs/blob/master/mkdocs/structure/files.py), is there some doc available on these objects? 

If there is a files collection available as argument for events, there must have been a design decision that this could have _some_ use? Could someone give more details (possible uses, how-to, etc.)? 

Ideally, could the documentation be expanded to cover that type of information?

**Comment by waylan** [comment]
This is the subject of #1629 (in part). The reorganization took place in #2344. Now that that is done, we need to begin documenting the API. Of course volunteers are welcome.

**Reply by fralau** [reply]
> Of course volunteers are welcome.

@waylan I cannot promise you anything, but I'm also a "doc guy". If you tell me how to get involved in that part, I might give it a thought. 🙂

**Reply by Andre601** [reply]
I could try to help improve the docs, since making documentation of any kind is a passion of mine.
But I'm not quite confident to start making a PR, simply because I don't know what has been done so far and what hasn't.
Also, I'm not quite sure about what the current formatting structure of the docs are because I noticed from my previous PR about the README, that there seems to be a set character limit per line and also some other formatting rules?

I feel like the "About" section could deserve a page about the formatting and contributing towards the docs themself.
Or the Contributing page could receive a section mentioning the formatting of docs.

**Reply by waylan** [reply]
As the last comment in #1629 mentions, we will be using [tomchristie/mkautodoc][1] for documenting the API. To start, the plugin needs to be enabled. The comments in the source code may need to be updated to more clearly define what parameters are accepted and what values are returned for each class/method/function. And finally, those comments need to be included in the documentation via mkautodoc's syntax at `dev-guide/API.md`

What has been done already is all in the `master` branch.

We should start by documenting the objects which get passed to the plugin events. We also need to document the config classes as plugins need those as well. Finally, some of the functions in `mkdocs.utils` may be useful to plugins.

Regarding the formatting, we are just using the standard [MarkdownLint rules][2], with a few [custom options][3]. If you write well-formed Markdown, you shouldn't have any issues with these rules.

[1]: https://github.com/tomchristie/mkautodoc
[2]: https://github.com/DavidAnson/markdownlint#rules--aliases
[3]: https://github.com/mkdocs/mkdocs/blob/master/.markdownlintrc

**Reply by oprypin** [reply]
https://github.com/tomchristie/mkautodoc is unmaintained. Is that decision still current?

**Reply by waylan** [reply]
> 
> 
> https://github.com/tomchristie/mkautodoc is unmaintained. Is that decision still current?

To be fair, I haven't used any of the various competing solutions. However, this one was implemented in the essentially same way I would have done it. And it was created by the creator of MkDocs. Those are the reasons why it was chosen.

Is there any other solutions listed in the wiki which take the same approach?

**Reply by fralau** [reply]
I am in favour of literate programing, but it should not become an impediment to documentation (and it is not strictly indispensable IMHO, since browsing code is no longer difficult, especially with github). 

I would favor the opposite approach, is to write introductions and summaries.

**Comment by fralau** [comment]
I understand the purpose of the tool [mkautodoc](https://github.com/tomchristie/mkautodoc), but there seems to be [mystery around the markup for the doctrings inside the Python source files](https://github.com/tomchristie/mkautodoc/issues/27).

It seems that a lot of prior knowledge is assumed for anyone who touches that subject (on the software stack, what tool does what, how this can be configured and parameterized, etc.)

Could you describe (in a nutshell) what would be the basic markup convention for docstrings you would be using, and where that convention is referenced?

**Reply by waylan** [reply]
It's been a while since I looked at it, but my recollection is that the answers to all those questions because clear once I looked at the code for the project. That said, I haven't actually used it, so perhaps not. I know, that's not particularly helpful. But unfortunately, I don't have time for more right now.

**Comment by pawamoy** [comment]
We are willing to support the default "mkdocs" theme in *mkdocstrings*, that is if maintainers consider using it instead of *mkautodoc*. We can discuss all this on our [Gitter room](https://gitter.im/mkdocstrings/community) or in this very discussion :slightly_smiling_face:

**Reply by waylan** [reply]
Of the various options available, *mkdocsstrings* is certainly a top contender and as it also injects into the Markdown document, is something I would be willing to use. It is also well documented and appears to be actively maintained.  Of course, the lack of support for the `mkdocs` theme is a show stopper for us. What would be involved in getting support added?

**Reply by pawamoy** [reply]
Bringing support for the `mkdocs` theme would consist in writing HTML templates and CSS files that integrate well with the theme. You would of course help us decide how each element should look like, such as signatures, arguments in docstrings, raised exceptions, etc.

What we can already do is render a random project's docs with the `mkdocs`  theme and see what's wrong / what looks bad.

**Reply by waylan** [reply]
As the `mkdocs` theme is built on bootstrap, I suppose the ideal solution would be to include the necessary hooks in the HTML templates so that bootstrap's CSS is utilized. However, there are inevitably going to be some tweaks needed, which would require custom CSS.

I suppose the first step then is to develop a list of the various types of content generated by *mkdocstrings* and which type of bootstrap container each should be put in.

**Reply by oprypin** [reply]
@waylan I am not familiar with Bootstrap CSS, so for now I just applied the exact elements that are being used on the MkDocs site itself right now. Namely, it's those definition lists.
See [commit `974ca90` (#272)](https://github.com/mkdocstrings/mkdocstrings/pull/272/commits/974ca9010efca1b8279767faf8efcd2470a8371d) which added initial support. It's easy to see how to apply different elements, if desired.

**Comment by fralau** [comment]
@waylan What would be the next steps on this issue? What would be expected of us, or somebody else?

**Reply by waylan** [reply]
Right now my time is focused elsewhere. Specifically I am trying to get the items in the 1.2 milestone completed so we can do that release. Until that is done, I don't have any time to dig into this. If anyone wants to help, I guess they can start by adding support for the MkDocs theme, whatever that entails (like I said, I haven't had time to look). If you need input from me to move forward on that, then that will have to wait until after 1.2 is released. However, if you can move forward without me, even better.

**Reply by fralau** [reply]
@waylan I understand your quandary.

I may not be the best person for working MkDocs theme.

But if you are OK, I might occasionally contribute to the doc by following my own approach: to add to the existing **user guides** and amend them, within the existing documentation framework (as as described in #1629 and  #2344).

IMHO the generation of **reference documentation** from docstrings might be kept as a distinct subject.

**Reply by oprypin** [reply]
I have converted an initial batch of items to mkdocstrings.
See https://github.com/oprypin/mkdocs/commit/docplug and especially the commit message.

The commit ([and documentation](https://mkdocstrings.github.io/handlers/python/#google-style)) will also answer your other question about docstring conventions.

Indeed, this does not significantly intersect with the act of *actually writing* documentation (other than merge conflicts 😋). Regardless of whether a docstring plugin is used *now*, you can already start structuring documentation with the expectation that some designated parts of it can be documented wholesale as a collection of methods. Or decide that that it's best to not do that.

**Reply by waylan** [reply]
Just a comment on the direction I would like to take this. API docs are not a replacement for prose, only a supplement to it. I realize the plugin events read as API docs, so I understand why you did this there. However, for the most part, the API docs should be in addition to what we have, not replace it. For example, the theme docs include documentation of the various variables available to a theme. That same information should also be repeated in the API docs. The difference is the the API docs simply document the Python object (it parameters, return values, attributes, methods, etc), whereas the theme docs document how those objects are used from a template. It would certainly be best if each object in the theme docs linked to the API docs for that item. Of course, the target audience of the API docs would be Python programers, whereas the target audience of the theme docs would be web developers.

In any event, yes, if you want to help with the documentation, a great way to start would be to work on writing the documentation as docstrings in the code.

**Reply by oprypin** [reply]
I am proceeding with mkdocstrings as mentioned - in https://github.com/mkdocs/mkdocs/pull/2931
